src/fsharp/FSharp.Core/printf.fs

```diff
@@ -1182,401 +935,288 @@ module internal PrintfImpl =
+                // identical results each time it is ok
+                functionFactory <- (funcObj :?> PrintfFuncFactory<'Printer, 'State, 'Residue, 'Result>)
+                stringCount <- (2 * count + 1)  
+            | _ -> ()
+            functionFactory
+
+    /// 2-level cache, keyed by format string and index types
+    type Cache<'Printer, 'State, 'Residue, 'Result>() =
+
+        /// 1st level cache (type-indexed). Stores last value that was consumed by the current thread in
+        /// thread-static field thus providing shortcuts for scenarios when printf is called in tight loop.
+        [<DefaultValue; ThreadStatic>]
+        static val mutable private mostRecent: FormatParser<'Printer, 'State, 'Residue, 'Result>
     
-        static member Get(key: Format<'T, 'State, 'Residue, 'Result>) =
-            if not (Cache<'T, 'State, 'Residue, 'Result>.last === null) 
-                && key.Value.Equals (fst Cache<'T, 'State, 'Residue, 'Result>.last) then
-                    snd Cache<'T, 'State, 'Residue, 'Result>.last
+        // 2nd level cache (type-indexed). Dictionary that maps format string to the corresponding cache entry
+        static let mutable dict : ConcurrentDictionary<string, FormatParser<'Printer, 'State, 'Residue, 'Result>> = null
+
+        static member GetParser(format: Format<'Printer, 'State, 'Residue, 'Result>) =
+            let cacheEntry = Cache<'Printer, 'State, 'Residue, 'Result>.mostRecent
+            let fmt = format.Value
+            if cacheEntry === null then 
+                let parser = FormatParser(fmt)
+                Cache.mostRecent <- parser
+                parser
+            elif fmt.Equals cacheEntry.FormatString then 
+                cacheEntry
             else
-                let v = get key.Value
-                Cache<'T, 'State, 'Residue, 'Result>.last <- (key.Value, v)
-                v
-
-    type StringPrintfEnv<'Result>(k, n) = 
+                // Initialize the 2nd level cache if necessary.  Note there's a race condition but it doesn't
+                // matter if we initialize these values twice (and lose one entry)
+                if isNull dict then 
+                    dict <- ConcurrentDictionary<_,_>()
+
+                let parser = 
+                    match dict.TryGetValue(fmt) with 
+                    | true, res -> res
+                    | _ -> 
+                        let parser = FormatParser(fmt)
+                        // There's a race condition - but the computation is functional and it doesn't matter if we do it twice
+                        dict.TryAdd(fmt, parser) |> ignore
+                        parser
+                Cache.mostRecent <- parser
+                parser
+
+    type LargeStringPrintfEnv<'Result>(continuation, blockSize) = 
         inherit PrintfEnv<unit, string, 'Result>(())
-
-        let buf: string[] = Array.zeroCreate n
+        let buf: string[] = Array.zeroCreate blockSize
```

Comment:
How large could this get? How likely is this array to get re-created if it's very large (e.g., 85k)?

Reply:
It gets created once on each format invocation, it holds the fragment of the formatting